Add becker covers

nicolasberthel opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The purpose of this change is to introduce a new component to control becker covers via becker centronic usb stick


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
becker
  device: "/dev/ttSY3"
  covers:
  - name: "Kitchen Cover"
    channel: "1"
  - name: "Bedroom Cover"
    channel: "2"
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

The component introduce as well some services in order to pair new covers with the stick (if not already paired before) : becker.pair

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [X] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
Add new component for becker covers integration using centronic usb stick

Commit:
Add new component for becker covers integration using centronic usb stick

Commit:
Change position of becker requirement

Commit:
Change position of becker requirement

Commit:
Add documentation to manifest (not yet published)

Commit:
Add documentation to manifest (not yet published)

homeassistant commented:
Hi @nicolasberthel, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#35288).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
Add missing new line

Commit:
Add missing new line

Commit:
Remove config flow and unused imports

Commit:
Remove config flow and unused imports

Commit:
Fix some code style

Commit:
Fix some code style

Commit:
Rephrase some comments

Commit:
Rephrase some comments

Commit:
Run isort to sort imports

Commit:
Run isort to sort imports

Commit:
re order some imports

Commit:
re order some imports

Commit:
Move voluptuous import

Commit:
Move voluptuous import

frenck commented:
Hi there @nicolasberthel!

Thanks for opening up a PR, that is appreciated.

However, we no longer accept integration that is set up via a platform configuration.

For more information see: <https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md>

Please note (not 100% sure), ADR-0010 might also apply here:

<https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md>

Commit:
Change Becker Component to allow Integration configuration via UI

Commit:
Change Becker Component to allow Integration configuration via UI

nicolasberthel commented:
Hi @frenck I changed the implementation to support integration via UI

Commit:
Fix some code style

Commit:
Fix some code style

Commit:
Fix some code style

Commit:
Fix some code style

nicolasberthel commented:
All checks looks to be green now, sorry it was pylint failing that seems not to run as pre commit

nicolasberthel commented:
I'm not sure what you expect from my side, I do not have news from the merge request but everything looks green on my side, so I'm wondering about this stale label

frenck commented on .coveragerc:
```diff
@@ -65,6 +65,7 @@ omit =
     homeassistant/components/azure_event_hub/*
     homeassistant/components/azure_service_bus/*
     homeassistant/components/baidu/tts.py
+    homeassistant/components/becker/*
```
The configuration flow should not be excluded from tests, by using * it is excluded.

frenck commented:
I'm sorry, @nicolasberthel. Busy times!

Some additional pointers before we can continue reviewing:
- The french translation file does not belong in this PR. Only the strings.json and the en.json. Please remove the fr.json file from this PR.
- The configuration flow requires tests with a 100% test coverage. Please add those tests.

frenck commented on homeassistant/components/becker/__init__.py:
```diff
@@ -0,0 +1,116 @@
+        hass.data[DOMAIN][CONF_COVERS][channel] = cover_conf
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data={"cover": cover_conf}
```
This is a new integration, there is nothing to import from previous integrations? I guess this can be removed?

nicolasberthel replied:
Hi @frenck thank you for your feedback, I've introduced unit tests and removed the french translation.

nicolasberthel replied:
This piece of code is to retrieve the covers from the configuration file (they need to be manually setup) the config flow is setting up the interface (usually the rf usb stick from becker)

balloob replied:
Ask the user to configure devices during the config flow or use an options flow to get this data from the users.

balloob replied:
I also see that there is a service too. Shouldn't that be enough ?

balloob replied:
@nicolasberthel could you answer this question please.

nicolasberthel replied:
@balloob  the service is "just" for the pairing (to pair the cover to the usb stick), the cover setup is not yet done via the UI, basically this is the list of covers that is still in the yaml file (i.e. documentation)

Commit:
Update comment

Commit:
Update comment

Commit:
move cover conf init to prevent empty data

Commit:
move cover conf init to prevent empty data

Commit:
remove french translation

Commit:
remove french translation

Commit:
fix some code style

Commit:
fix some code style

Commit:
add test requirements

Commit:
add test requirements

Commit:
Improve test coverage

Commit:
Improve test coverage

nicolasberthel commented:
Hi @frenck I can imagine that you are still quite busy :) Are you still missing something from me here ?
Thanks 
Nicolas

frenck commented on homeassistant/components/becker/cover.py:
```diff
@@ -0,0 +1,201 @@
+
+    async def async_update(self):
+        """Update the position of the cover."""
+        if self._template is not None:
```
Turn this into a guard, so the rest of the function doesn't need to be indented.
```suggestion
        if self._template is None:
            return
```

frenck commented on homeassistant/components/becker/cover.py:
```diff
@@ -0,0 +1,201 @@
+    }
+)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
Platform schemas are no longer allowed as per ADR-0007

See: <https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md#decision>

Commit:
Update homeassistant/components/becker/__init__.py

Update for loop

Co-authored-by: Franck Nijhof <git@frenck.dev>

Commit:
Update homeassistant/components/becker/__init__.py

Update for loop

Co-authored-by: Franck Nijhof <git@frenck.dev>

Commit:
Update homeassistant/components/becker/__init__.py

change way of accessing data

Co-authored-by: Franck Nijhof <git@frenck.dev>

Commit:
Update homeassistant/components/becker/__init__.py

change way of accessing data

Co-authored-by: Franck Nijhof <git@frenck.dev>

nicolasberthel commented on homeassistant/components/becker/__init__.py:
```diff
@@ -0,0 +1,116 @@
+    if not becker:
+        return False
+
+    for _ in range(0, 2):
```
Yes this was a finding of how becker devices are working. When calling a stop command twice you are sure to initialize an internal counter correctly and to be able to communicate with the covers.

Commit:
Update homeassistant/components/becker/config_flow.py

Fix class definition

Co-authored-by: Franck Nijhof <git@frenck.dev>

Commit:
Update homeassistant/components/becker/config_flow.py

Fix class definition

Co-authored-by: Franck Nijhof <git@frenck.dev>

Commit:
Remove upper case in debug message

Commit:
Remove upper case in debug message

Commit:
Implement ConfigEntryNotReady Exception

Commit:
Implement ConfigEntryNotReady Exception

Commit:
ignore unsued import in pylint

Commit:
ignore unsued import in pylint

Commit:
Define config schema inline

Co-authored-by: Franck Nijhof <git@frenck.dev>

Commit:
Define config schema inline

Co-authored-by: Franck Nijhof <git@frenck.dev>

Commit:
remove platform schema

Commit:
remove platform schema

Commit:
Add device path

Commit:
Add device path

Commit:
Change config flow to provide a list of available ports

Commit:
Change config flow to provide a list of available ports

Commit:
support channels with multiple units

Commit:
support channels with multiple units

nicolasberthel commented on homeassistant/components/becker/__init__.py:
```diff
@@ -0,0 +1,116 @@
+        stick_path = hass.data[DOMAIN][CONF_DEVICE]
+
+    _LOGGER.debug("SETTING CENTRONIC STICK ON PORT {stick_path}")
+    becker = BeckerConnection(stick_path)
```
@frenck now it will, what do you suggest in case of exception here ?

Commit:
Use latest pybecker lib

Commit:
Use latest pybecker lib

Commit:
Handle exception when connection to becker centronic stick

Commit:
Handle exception when connection to becker centronic stick

Commit:
Change config flow to display list of available com ports

Commit:
Change config flow to display list of available com ports

nicolasberthel commented on homeassistant/components/becker/config_flow.py:
```diff
@@ -0,0 +1,47 @@
+            return self.async_abort(reason="one_instance_only")
+
+        fields = {}
+        fields[vol.Required(CONF_DEVICE, default=DEFAULT_CONF_USB_STICK_PATH)] = str
```
Thank you for the hint, I made a change in this direction

Commit:
fix some code style

Commit:
fix some code style

frenck commented:
not stale

benganellison commented:
I'm looking for an integration for becker. is there a way that I can help out here?

benganellison commented:
@frenck there is a reply to you, in the change you asked for, that you have not answered. can you see if you can help out a bit? please!

frenck commented on homeassistant/components/becker/__init__.py:
```diff
@@ -0,0 +1,124 @@
+        becker = BeckerConnection(stick_path)
+    except BeckerConnectionError:
+        _LOGGER.error("Fail to connect to becker device")
+    except Exception as error:
+        _LOGGER.error("Fail to connect to becker device %s", error)
```
We should not catch a broad exception like this.

nicolasberthel replied:
You a right and I even think this make no sense because the only exception that should occur here is the BeckerConnectionError (that is already caught), I simply dropped this broad catch

frenck commented on homeassistant/components/becker/__init__.py:
```diff
@@ -0,0 +1,124 @@
+
+    try:
+        becker = BeckerConnection(stick_path)
+    except BeckerConnectionError:
```
Should this caught exception throw a ConfigEntryNotReady?

nicolasberthel replied:
Yes this is a good Idea I added a Raise of ConfigEntryNotReady

Commit:
Update __init__.py

Raise ConfigEntryNotReady in case we Get a BeckerConnectionError when trying to setup becker usb device

Commit:
Update __init__.py

Raise ConfigEntryNotReady in case we Get a BeckerConnectionError when trying to setup becker usb device

Commit:
Merge branch 'dev' into becker-integration

Commit:
Merge remote-tracking branch 'origin/becker-integration' into becker-integration

Commit:
Merge remote-tracking branch 'origin/becker-integration' into becker-integration

homeassistant commented:
Hello @nicolasberthel,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`c0171368f9428a980a0a25c5de71c495e3c631c5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;35288&#x2F;commits&#x2F;c0171368f9428a980a0a25c5de71c495e3c631c5) - This commit has something that looks like an email address (nicolas.berthel@enovos.eu). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

Commit:
Update __init__.py

Add from which exception the ConfigEntryNotReady is raised

homeassistant commented:
Hello @nicolasberthel,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`c0171368f9428a980a0a25c5de71c495e3c631c5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;35288&#x2F;commits&#x2F;c0171368f9428a980a0a25c5de71c495e3c631c5) - This commit has something that looks like an email address (nicolas.berthel@enovos.eu). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

Commit:
Update config_flows.py

Regenerate config_flow

homeassistant commented:
Hello @nicolasberthel,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`c0171368f9428a980a0a25c5de71c495e3c631c5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;35288&#x2F;commits&#x2F;c0171368f9428a980a0a25c5de71c495e3c631c5) - This commit has something that looks like an email address (nicolas.berthel@enovos.eu). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

Commit:
regenerate config flow

homeassistant commented:
Hello @nicolasberthel,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`c0171368f9428a980a0a25c5de71c495e3c631c5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;35288&#x2F;commits&#x2F;c0171368f9428a980a0a25c5de71c495e3c631c5) - This commit has something that looks like an email address (nicolas.berthel@enovos.eu). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

Commit:
fix code style

homeassistant commented:
Hello @nicolasberthel,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`c0171368f9428a980a0a25c5de71c495e3c631c5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;35288&#x2F;commits&#x2F;c0171368f9428a980a0a25c5de71c495e3c631c5) - This commit has something that looks like an email address (nicolas.berthel@enovos.eu). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

Commit:
rollback requirement

MartinHjelmare commented:
Please fix the CLA issues with your commits. See the bot comments.

homeassistant commented:
Hello @nicolasberthel,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`c0171368f9428a980a0a25c5de71c495e3c631c5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;35288&#x2F;commits&#x2F;c0171368f9428a980a0a25c5de71c495e3c631c5) - This commit has something that looks like an email address (nicolas.berthel@enovos.eu). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

balloob commented:
Closing this as CLA hasn't been signed. Ping us if you've signed the CLA.

Commit:
Merge remote-tracking branch 'origin/becker-integration' into becker-integration

nicolasberthel commented:
Hello, sorry for this cla issue, I signed it finally (always thought it was done) 
I also added my other email to my gitlab account
So this part should not be blocking any longer
For the rest of the open point I think they are mainly addressed
With my latest merge I see that I get some code check failures (with lint but didn't yet figure out why)

Commit:
Merge remote-tracking branch 'origin/becker-integration' into becker-integration

nicolasberthel commented:
CC @balloob

balloob commented on homeassistant/components/becker/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+        schema = vol.Schema({vol.Required(CONF_DEVICE_PATH): vol.In(list_of_ports)})
+        return self.async_show_form(step_id="user", data_schema=schema)
+
+    async def async_step_import(self, info):
```
We shouldn't support importing. It's a new integration and so this is not allowed.

balloob replied:
This is something that would still need addressing. Why can't you use the service ?

nicolasberthel replied:
You're right, I'll remove this part, this was basically for my own case as I was using this a custom component

nicolasberthel replied:
@balloob finally it was really used as the covers are setup via yaml, this import is used to init the entities in async_setup
maybe it's not the right way but I was asked so many time to change the implementation that I'm completely lost. To be honnest it's working fine for me as a custom component, it's sad but I think I'll keep it for me as it is then

benganellison replied:
this is too bad, been waiting for this since first pushed in may. I will look in to your code nicolasberthel to see if i atleast can have it playing nice in my setup as a custom component

nicolasberthel replied:
> this is too bad, been waiting for this since first pushed in may. I will look in to your code nicolasberthel to see if i atleast can have it playing nice in my setup as a custom component
If you are more confident with Python that I am feel free to use my code, I can handover the integration to you ;)

benganellison replied:
I'm definitely not confident at all. Can barely do anything

balloob commented on homeassistant/components/becker/__init__.py:
```diff
@@ -0,0 +1,122 @@
+    _LOGGER.info("Setting Centronic stick on port %s", stick_path)
+
+    try:
+        becker = BeckerConnection(stick_path)
```
Given that this can raise a connection error, that means that the `BeckerConnection` constructor does I/O? In that case it should be run inside the executor:
```suggestion
        becker = await hass.async_add_executor_job(BeckerConnection, stick_path)
```

balloob commented on homeassistant/components/becker/__init__.py:
```diff
@@ -0,0 +1,122 @@
+
+    hass.data[DOMAIN]["connector"] = becker.connection
+
+    hass.data.setdefault(DOMAIN, {}).update({entry.entry_id: becker})
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = becker
```

balloob commented on homeassistant/components/becker/__init__.py:
```diff
@@ -0,0 +1,122 @@
+    )
+
+    hass.services.async_register(DOMAIN, "pair", becker.handle_pair, PAIR_SCHEMA)
+    hass.services.async_register(DOMAIN, "log_units", becker.handle_log_units)
```
This service has no schema?

nicolasberthel replied:
yes it is just logging the available units with the details (mostly the internal ID an counter), I think it would make sense to register it as well as admin service BTW

balloob commented on homeassistant/components/becker/__init__.py:
```diff
@@ -0,0 +1,122 @@
+        hass.config_entries.async_forward_entry_setup(entry, "cover")
+    )
+
+    hass.services.async_register(DOMAIN, "pair", becker.handle_pair, PAIR_SCHEMA)
```
Use `async_register_admin_service` from `homeassistant.helpers.service` to ensure that only admin users can manage pairing.

balloob commented on homeassistant/components/becker/cover.py:
```diff
@@ -0,0 +1,191 @@
+
+_VALID_STATES = [STATE_OPEN, STATE_CLOSED, "true", "false"]
+
+COVER_SCHEMA = vol.Schema(
```
This is not used.

nicolasberthel replied:
I think it's used in 
`if state in _VALID_STATES:` (line 177)

nicolasberthel replied:
os tempate support is removed I as well removed the _VALID_STATES

balloob commented on homeassistant/components/becker/cover.py:
```diff
@@ -0,0 +1,191 @@
+            _LOGGER.error("Must specify {CONF_CHANNEL}")
+            continue
+        templates = {
+            CONF_VALUE_TEMPLATE: state_template,
```
Can you explain what is going on here? 

A user should not be required to implement their own templates.

nicolasberthel replied:
This is used if you want to template the state of a cover (for instance with a open/close sensor)

balloob replied:
That's not something we should support inside the integration that represents the data from the API. Such features need to be added in a template sensor that forwards commands to this entity.

nicolasberthel replied:
Ok I've remove everything about templates for cover positions then

Commit:
Only allow pairing when registered as admin user

Commit:
Update homeassistant/components/becker/__init__.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

Commit:
Update homeassistant/components/becker/__init__.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

Commit:
Merge branch 'becker-integration' of github.com:nicolasberthel/core into becker-integration

Commit:
Remove template support for cover position

Commit:
fix some code style

Commit:
fix some code style

Commit:
remove asynctest

Commit:
fix test case

Commit:
Remove import feature

Commit:
Remove import test

Commit:
Remove import test

Commit:
Re enable import flow

milanmeu commented on homeassistant/components/becker/manifest.json:
```diff
@@ -0,0 +1,9 @@
+    "name": "Becker",
+    "documentation": "https://www.home-assistant.io/integrations/becker",
+    "requirements": ["pybecker==1.0.4"],
+    "dependencies": [],
```
Remove empty
```suggestion
```

milanmeu commented on homeassistant/components/becker/translations/en.json:
```diff
@@ -0,0 +1,17 @@
+{
```
This file can be automatically generated by this script:
```bash
python3 -m script.translations develop
```
It prevents errors such as a blank line at the end, tabs instead of spaces, etc...

milanmeu reviewed (changes requested):
The `strings.json` file is missing.
Developers docs: https://developers.home-assistant.io/docs/internationalization/core

cgarwood commented:
PR seems to have gone stale. Feel free to submit a new PR when you're ready to finish 👍
